### DISCUSSION TITLE
Append tags to Mailchimp contact instead of overwriting

### DISCUSSION THREAD

**Discussion by jimirobaer** [solspace/craft-freeform#547] [Feature Requests]
**Describe the bug or issue you're experiencing**

We recently moved from ActiveCampaign to Mailchimp for our email marketing needs and are switching our Freeform forms to use Mailchimp as well. We noticed a difference in the behaviour of Freeform regarding tags you can assign to contacts.

In ActiveCampaign, tags from Freeform were added to existing tags. But in Mailchimp, pushing tags from Freeform seems to completely replace the existing tags. 

It is not default behaviour for the Mailchimp API to overwrite all tags if you pass new ones throught a POST request. I can confirm this by manually passing through a new tag using a POST request via Postman. Tags are appended to existing tags for a contact. I used the URLs as described in their own [Mailchimp API docs](https://mailchimp.com/developer/marketing/api/list-members/). 

POST request URL used to post tag(s): `https://xxx.api.mailchimp.com/3.0/lists/{list_id}/members/{subscriber_hash}/tags`.

**Steps to reproduce**

1. Add Mailchimp integration in Email Marketing settings and add subscriber field to form.
2. Create empty field with default value (eg. "Test") and setup this field as contents for Mailchimp Tags.
3. Push to Mailchimp on submit.

**Expected behavior**

Tags being appended to Mailchimp, not replace existing tags.

**Craft & Plugin Info (please complete the following information):**

- Craft Version: 3.5.16
- Freeform Version: 3.13.22
- Freeform Edition: Pro
- Fresh Install or Upgrade: Upgrade

**Comment by kjmartens** [comment]
Hi @jimirobaer,

We appreciate your feedback. However, some customers also rely on it to replace tags instead of append tags. In order to account for all use-cases, we would need to add a setting. I've made a note about this for future consideration. 🙂

**Reply by jwmatlock** [reply]
@kjmartens Has this always been the case? Trying to determine for how long our client's form contacts tags have been getting cleared out in Mailchimp. This was unexpected as the default behavior for Mailchimp tags is appending, not replacing.

**Comment by jimirobaer** [comment]
Sure, I get that. 

We have been working on adding on our layer of talking with Mailchimp on top of the forms. This allows us to better control whatever is sent to Mailchimp. We are working with the official Mailchimp Marketing API for PHP and sending responses to Mailchimp on submit using Freeform's `SubmitEvent`.

However, we are still running into the issue that Freeform is resetting our tags. This is the current chain of events on submitting a form:

1. The Mailchimp integration in Freeform is triggered (which has all the field mapping) and sends response to Mailchimp.
2. Our own methods to send additional information (tags, events) to the Mailchimp endpoints.

Because we left the "Tags" field in the CMS open, I suspect Freeform is still resetting them. I think that's unusual behaviour as there's is no endpoint as far as I know that simply '_resets_' all tags.

We've been stuck on this for a few days and I'm not exactly sure what to do here other than handle the Mailchimp submission entirely myself, which I'm not really excited about.

**Reply by jwmatlock** [reply]
@jimirobaer Did you come up with a solution or did you abandon Freeform because of this issue? We just now realized this was happening, where tags were being reset. Not expected behavior and has caused major problems.

**Reply by jimirobaer** [reply]
@jwmatlock Sorry for the late reply. We kept Freeform but went with a custom handler for the tags. We built it around the official Mailchimp PHP library.

**Comment by jwmatlock** [comment]
We are also just realizing this issue. The default behavior for Mailchimp tags should be that they are appended, not replaced. Is there any way to work around this current issue?

**Comment by kjmartens** [comment]
I've queued this up to be addressed in an upcoming version of Freeform (likely within 2 weeks). We'll add a setting inside the MailChimp integration that allows you to choose the behavior. Thanks again for your feedback @jimirobaer and @jwmatlock 🙂

**Reply by jwmatlock** [reply]
Thank you!

**Comment by kjmartens** [comment]
This is now possible in Freeform 4.0.24+ via a new setting inside the Mailchimp integration settings. 🙂
